lib/vrl/parser/src/parser.lalrpop

```diff
@@ -0,0 +1,501 @@
+
+AssignmentExpr: Node<Expr> = {
+    Assignment => Node::new(<>.span(), Expr::Assignment(<>)),
+    ArithmeticExpr,
+};
+
+Assignment: Node<Assignment> = {
+    Sp<AssignmentSingle>,
+    Sp<AssignmentInfallible>,
+};
+
+AssignmentSingle: Assignment = {
+    <target: Sp<AssignmentTarget>> "="
+        <expr: Box<Expr>> => Assignment::Single { target, expr },
+}
+
+AssignmentInfallible: Assignment = {
+    <ok: Sp<AssignmentTarget>> ","
+        <err: Sp<AssignmentTarget>> "="
+        <expr: Box<Expr>> => Assignment::Infallible{ ok, err, expr},
+}
+
+#[inline]
+AssignmentTarget: AssignmentTarget = {
+    "_" => AssignmentTarget::Noop,
+    "." <Path?> => AssignmentTarget::External(<>),
+    Ident Path? => AssignmentTarget::Internal(<>),
+};
+
+// -----------------------------------------------------------------------------
+// arithmetic
+// -----------------------------------------------------------------------------
+
+ArithmeticExpr: Node<Expr> = {
+    Sp<ErrorCoalesce>,
+};
+
+ErrorCoalesce: Expr = {
+    Op<ErrorCoalesce, "??", Logical>,
+    Logical,
+};
+
+Logical: Expr = {
+    Op<Logical, "||", Equal>,
+    Op<Logical, "&&", Equal>,
+    Equal,
+};
+
+Equal: Expr = {
+    Op<Equal, "!=", Compare>,
+    Op<Equal, "==", Compare>,
+    Compare,
+};
+
+Compare: Expr = {
+    Op<Compare, ">=", Add>,
+    Op<Compare, ">", Add>,
+    Op<Compare, "<=", Add>,
+    Op<Compare, "<", Add>,
+    Add,
+};
+
+Add: Expr = {
+    Op<Add, "+", Factor>,
+    Op<Add, "-", Factor>,
+    Factor,
+};
+
+Factor: Expr = {
+    Op<Factor, "%", Term>,
+    Op<Factor, "*", Term>,
+    Op<Factor, "/", Term>,
+    Term,
+};
+
+Term: Expr = {
+    Sp<Literal> => Expr::Literal(<>),
+    Sp<Container> => Expr::Container(<>),
+    Sp<Query> => Expr::Query(<>),
+    Sp<FunctionCall> => Expr::FunctionCall(<>),
+    Sp<Ident> => Expr::Variable(<>),
+};
+
+// -----------------------------------------------------------------------------
+// query
+// -----------------------------------------------------------------------------
+
+pub Query: Query = {
+    LQuery <Sp<".">> RQuery => {
+        let span = <>.span();
+        let target = Node::new(span, QueryTarget::External);
+        let path = Node::new(span, Path(vec![]));
+
+        Query { target, path }
+    },
+    LQuery <target: Sp<QueryTarget>> <path: Sp<Path>> RQuery => Query { target, path },
+};
+
+QueryTarget: QueryTarget = {
+    Ident => QueryTarget::Internal(<>),
+    "." => QueryTarget::External,
+    FunctionCall => QueryTarget::FunctionCall(<>),
+    Container => QueryTarget::Container(<>),
+};
+
+// -----------------------------------------------------------------------------
+// path
+// -----------------------------------------------------------------------------
+
+Path: Path = PathSegment+ => Path(<>);
+
+PathSegment: PathSegment = {
+    "."? <Field> => PathSegment::Field(<>),
+    "[" <Integer> "]" => PathSegment::Index(<>),
```

Comment:
It would be handy (at some point) to allow expressions as well as Integers here..

Reply:
Yeah. I've thought about it, but I'm hesitant. I worry that it'll add significant complexity to our path handling if we start making them dynamic (the next request would be to allow dynamic fields).

Reply:
I'm going to mark this as "resolved", as we're not going to tackle this in this PR (and I suspect not in the near future either).